File: services/python-worker/plugins/base_plugin.py
"""
Base Plugin Interface
All plugins must inherit from this class
"""

from abc import ABC, abstractmethod
from typing import Dict, Any, Optional


class BasePlugin(ABC):
    """
    Base class for all worker plugins
    
    Each plugin represents one capability that the worker can perform.
    Simply create a new .py file in the plugins/ directory and define a class
    that inherits from BasePlugin.
    """
    
    @property
    @abstractmethod
    def name(self) -> str:
        """Capability name (e.g., 'hello', 'process_image')"""
        pass
    
    @property
    @abstractmethod
    def description(self) -> str:
        """Human-readable description of what this plugin does"""
        pass
    
    @property
    def input_schema(self) -> str:
        """JSON schema for input validation (optional)"""
        return "{}"
    
    @property
    def output_schema(self) -> str:
        """JSON schema for output format (optional)"""
        return "{}"
    
    @property
    def http_method(self) -> str:
        """HTTP method for Web API endpoint (GET/POST/PUT/DELETE)"""
        return "POST"
    
    @property
    def accepts_file(self) -> bool:
        """Whether this capability accepts file upload"""
        return False
    
    @property
    def file_field_name(self) -> Optional[str]:
        """Field name for file upload (if accepts_file is True)"""
        return None
    
    @abstractmethod
    def execute(self, params: Dict[str, Any], worker_sdk=None) -> Dict[str, Any]:
        """
        Execute the plugin logic
        
        Args:
            params: Input parameters from the request
            worker_sdk: Reference to worker SDK for calling other workers
            
        Returns:
            Dict with the result data
        """
        pass
    
    def on_load(self):
        """Called when plugin is loaded (optional hook)"""
        pass
    
    def on_unload(self):
        """Called when plugin is unloaded (optional hook)"""
        pass

File: services/java-simple-worker/src/main/java/com/deepapp/worker/plugins/BasePlugin.java
package com.deepapp.worker.plugins;

import java.util.Map;

/**
 * Base interface for all worker plugins
 * 
 * Each plugin represents one capability that the worker can perform.
 * Simply create a new class in the plugins package that implements this interface.
 */
public interface BasePlugin {
    
    /**
     * Get the capability name (e.g., "hello_world", "process_data")
     */
    String getName();
    
    /**
     * Get human-readable description
     */
    String getDescription();
    
    /**
     * Get JSON schema for input validation (optional)
     */
    default String getInputSchema() {
        return "{}";
    }
    
    /**
     * Get JSON schema for output format (optional)
     */
    default String getOutputSchema() {
        return "{}";
    }
    
    /**
     * Get HTTP method for Web API endpoint
     */
    default String getHttpMethod() {
        return "POST";
    }
    
    /**
     * Whether this capability accepts file upload
     */
    default boolean acceptsFile() {
        return false;
    }
    
    /**
     * Execute the plugin logic
     * 
     * @param input Input parameters as JSON string
     * @param workerSDK Reference to worker SDK for calling other workers
     * @return Result as JSON string
     * @throws Exception if execution fails
     */
    String execute(String input, Object workerSDK) throws Exception;
    
    /**
     * Called when plugin is loaded (optional hook)
     */
    default void onLoad() {
        // Override if needed
    }
    
    /**
     * Called when plugin is unloaded (optional hook)
     */
    default void onUnload() {
        // Override if needed
    }
}
